CMakeLists.txt

```diff
@@ -0,0 +1,474 @@
+
+        add_custom_command(
+            OUTPUT ${CUR_OUTPUT}
+            COMMAND cp ${CUR_INPUT} build/${CUR_INPUT} && protoc -I. --cpp_out=. --python_out=. build/${CUR_INPUT}
+            DEPENDS ${CUR_INPUT} ${OUTPUT_PATH}
+            )
+    endforeach()
+
+endmacro(c_proto_buff)
+
+c_proto_buff(CORE_PROTO CORE_PROTO_SRC)
+c_proto_buff(PROTOCOL_PROTO PROTOCOL_SRC)
+c_proto_buff(TREE_PROTO TREE_PROTO_SRC)
+c_proto_buff(SERVER_PROTO SERVER_PROTO_SRC)
+c_proto_buff(STORAGE_PROTO STORAGE_PROTO_SRC)
+
+## append compiler flags (and generate a object)
+
+# logcabin srcs
+set_property(
+    SOURCE
+    Storage/Tool.cc
+    Server/Main.cc
+    ${CORE_SRC}
+    ${CLIENT_SRC}
+    ${EVENT_SRC}
+    ${RPC_SRC}
+    ${SERVER_SRC}
+    ${STORAGE_SRC}
+    ${TREE_SRC}
+    ${EXAMPLE_FILES}
+    APPEND_STRING
+    PROPERTY
+    COMPILE_FLAGS
+    " ${CMAKE_LOGCABIN_CXX_FLAGS}"
+    )
+
+# protobuf flags
+set_property(
+    SOURCE
+    ${CORE_PROTO_SRC}
+    ${PROTOCOL_SRC}
+    ${TREE_PROTO_SRC}
+    ${SERVER_PROTO_SRC}
+    ${STORAGE_PROTO_SRC}
+    APPEND_STRING
+    PROPERTY
+    COMPILE_FLAGS
+    " ${CMAKE_PROTO_CXX_FLAGS}"
+    )
+
+set_property(
+    SOURCE
+    gtest/src/gtest-all.cc
+    test/TestRunner.cc
+    ${TEST_FILES}
+    ${MOCK_FILES}
+    APPEND_STRING
+    PROPERTY
+    COMPILE_FLAGS
+    # -fno-access-control allows tests to access private members
+    " ${CMAKE_GTEST_CXX_FLAGS} ${CMAKE_LOGCABIN_CXX_FLAGS} -fno-access-control"
+    )
+
+##executable daemon
+add_executable(LogCabin
+    Server/Main.cc
+    )
+
+add_executable(ServerControl
+    Client/ServerControl.cc
+    )
+
+add_library(logcabin STATIC 
+    ${CLIENT_SRC}
+    ${TREE_SRC}
+    ${TREE_PROTO_SRC}
+    ${PROTOCOL_SRC}
+    ${RPC_SRC}
+    ${EVENT_SRC}
+    ${CORE_SRC}
+    ${CORE_PROTO_SRC}
+    )
+
+add_library(ServerSrcObject STATIC 
+    ${SERVER_SRC}
+    ${SERVER_PROTO_SRC}
+    )
+
+add_library(StorageSrcObject STATIC 
+    ${STORAGE_PROTO_SRC}
+    ${STORAGE_SRC}
+    )
+
+add_executable(StorageTool
+    Storage/Tool.cc
+    )
+
+add_custom_target(check
+    COMMAND scripts/cpplint.py
+    )
+add_custom_target(lint
+    COMMAND scripts/cpplint.py
+    )
+add_custom_target(docs
+    COMMAND doxygen docs/Doxyfile
+    )
+add_custom_target(tags
+    COMMAND ctags -R --exclude=build --exclude=docs --exclude=CMakeFiles .
+    )
+
+# check library exists
+function(check_lib_exist LIB_NAME)
+    set(PATH_VAR_NAME "${LIB_NAME}_PATH")
+    find_library(${PATH_VAR_NAME} NAMES ${LIB_NAME})
+    if("${${PATH_VAR_NAME}}" STREQUAL "${PATH_VAR_NAME}-NOTFOUND")
+        ## sending warrning alows user to compile the lib without unnecessary dependency 
+        message(WARNING "can not find library ${LIB_NAME}")
+    endif()
+endfunction(check_lib_exist)
+
+check_lib_exist(pthread)
+check_lib_exist(protobuf)
+check_lib_exist(rt)
+check_lib_exist(cryptopp)
+
+# collect all libs when they all exist
+SET(LIBRERYS pthread protobuf rt cryptopp)
```

Comment:
LIBRARYS is closer but still not correct. The plural of library is libraries. Programmers often shorten it to libs, so pick one of those two please.